scipy/stats/_hypotests.py

```diff
@@ -9,6 +9,223 @@
+corresponding element of the output.
+If ``None``, the input will be raveled before computing the statistic."""
+.split('\n'))
+_axis_parameter_doc = Parameter(_name, _type, _desc)
+_axis_parameter = inspect.Parameter(_name,
+                                    inspect.Parameter.KEYWORD_ONLY,
+                                    default=0)
+
+_name = 'nan_policy'
+_type = "{'propagate', 'omit', 'raise'}"
+_desc = (
+"""Defines how to handle input NaNs.
+
+- ``propagate``: if a NaN is present in the axis slice (e.g. row) along
+  which the  statistic is computed, the corresponding entry of the output
+  will be NaN.
+- ``omit``: NaNs will be omitted when performing the calculation.
+  If insufficient data remains in the axis slice along which the
+  statistic is computed, the corresponding entry of the output will be
+  NaN.
+- ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
+.split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _vectorize_hypotest_factory(result_object, default_axis=0,
+                                n_samples=1, paired=False,
+                                result_unpacker=None, too_small=0,
+                                nan_policy_position=None, vectorized=False):
+
+    if result_unpacker is None:
+        def result_unpacker(res):
+            return res[..., 0], res[..., 1]
+
+    def is_too_small(samples):
+        for sample in samples:
+            if len(sample) <= too_small:
+                return True
+        return False
+
+    def vectorize_hypotest_decorator(hypotest_fun_in):
+        @wraps(hypotest_fun_in)
+        def vectorize_hypotest_wrapper(*args, axis=default_axis,
+                                       nan_policy='propagate', _no_deco=False,
+                                       **kwds):
+
+            if _no_deco:  # for testing, decorator does nothing
+                return hypotest_fun_in(*args, **kwds)
+
+            if (nan_policy_position is not None
+                    and len(args) > nan_policy_position):
+                # This is for overriding an existing nan_policy implementation.
+                # Consider what to do if both are specified but not the same.
+                # To detect this, should the default value of keyword arg be
+                # None?
+                nan_policy = args[nan_policy_position]
+
+            # if n_samples is None, all args are samples
+            n_samp = len(args) if n_samples is None else n_samples
+
+            # split samples from other arguments
+            samples = [np.atleast_1d(sample) for sample in args[:n_samp]]
+            args = args[n_samp:]
+
+            if axis is None:
+                samples = [sample.ravel() for sample in samples]
+                axis = 0
```

Comment:
To generalize to axis tuples, something like:
```suggestion
            if axis is None:
                samples = [sample.ravel() for sample in samples]
            else:
                # make sure axis ints are turned into a tuple in input validations
                # use _broadcast_array_shapes with np.broadcast_to to broadcast the samples
                samples = [np.movedim(sample, axis, range(-len(axis), 0) for sample in samples]
                samples = [sample.reshape(sample.shape[:-len(axis)] + (-1,)) for sample in samples]
            axis = -1
```